graph/src/data/subgraph/mod.rs

```diff
@@ -451,37 +451,53 @@ pub struct Graft {
     pub block: BlockNumber,
 }
 
+macro_rules! gbi {
+    ( $( $x:expr ),* ) => {
+        {
+            vec![
+                $(
+                    SubgraphManifestValidationError::GraftBaseInvalid($x),
+                )*
+            ]
+        }
+    };
+}
+
 impl Graft {
     async fn validate<S: SubgraphStore>(
         &self,
         store: Arc<S>,
     ) -> Vec<SubgraphManifestValidationError> {
-        fn gbi(msg: String) -> Vec<SubgraphManifestValidationError> {
-            vec![SubgraphManifestValidationError::GraftBaseInvalid(msg)]
-        }
-
         // We are being defensive here: we don't know which specific
         // instance of a subgraph we will use as the base for the graft,
         // since the notion of which of these instances is active can change
         // between this check and when the graft actually happens when the
         // subgraph is started. We therefore check that any instance of the
         // base subgraph is suitable.
-        match store.least_block_ptr(&self.base).await {
-            Err(e) => gbi(e.to_string()),
-            Ok(None) => gbi(format!(
+        match (
+            store.least_block_ptr(&self.base).await,
+            store.is_healthy(&self.base).await,
+        ) {
```

Comment:
If I understand correctly, isn't this disallowing grafts for any failed subgraph?

Reply:
No, take a look at the lines below:

```rust
// Base didn't process any block: graft should fail.
(Ok(None), _) => /* ... */,
// Base is healthy, however graft must be prior to processed block.
(Ok(Some(ptr)), Ok(true)) if ptr.number < self.block => /* ... */,
// Base failed, it can be grafted, but only for blocks prior to the failure.
(Ok(Some(ptr)), Ok(false)) if !(self.block < ptr.number) => /* ... */,
```

Reply:
Also, there's a test I made to so we can be sure failed subgraphs can be grafted prior to the failed block 🙂 
https://github.com/graphprotocol/graph-node/pull/3627/files#diff-8c7c963bca4ee64c86ba1963ed4c5f86dc40f89ccd3e890e4c508ba262e413f3R191